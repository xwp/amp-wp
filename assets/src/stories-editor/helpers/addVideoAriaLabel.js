/**
 * External dependencies
 */
import { ReactElement } from 'react';

/**
 * WordPress dependencies
 */
import { cloneElement } from '@wordpress/element';

/**
 * Helper to add an `aria-label` to video elements when saved.
 *
 * This helper is designed as a filter for `blocks.getSaveElement`.
 *
 * @param {ReactElement}  element     Previously generated React element
 * @param {Object}        type        Block type definition.
 * @param {string}        type.name   Name of block type
 * @param {Object}        attributes  Block attributes.
 *
 * @return {ReactElement}  New React element
 */
const addVideoAriaLabel = ( element, { name }, attributes ) => {
	// this filter only applies to core video objects (which always has children) where an aria label has been set
	if ( name !== 'core/video' || ! element.props.children || ! attributes.ampAriaLabel ) {
		return element;
	}

	/* `element` will be a react structure like:

	<figure>
		<amp-video|video>
			Fallback content
		</amp-video|video>
		[<figcaption>Caption</figcaption>]
	</figure>

	The video element can be either an `<amp-video>`` or a regular `<video>`.

	`<figcaption>` is not necessarily present.

	We need to hook into this element and add an `aria-label` on the `<amp-video|video>` element.
	*/

	const isFigure = element.type === 'figure';
	const childNodes = Array.isArray( element.props.children ) ? element.props.children : [ element.props.children ];
	const videoTypes = [ 'amp-video', 'video' ];
	const isFirstChildVideoType = videoTypes.includes( childNodes[ 0 ].type );
	if ( ! isFigure || ! isFirstChildVideoType ) {
		return element;
	}

	const figure = element;
	const [ video, ...rest ] = childNodes;

	const newVideo = cloneElement(
		video,
		{ 'aria-label': attributes.ampAriaLabel },
		video.props.children,
	);

	return cloneElement(
		figure,
		{},
		newVideo,
		...rest,
	);
};

export default addVideoAriaLabel;
