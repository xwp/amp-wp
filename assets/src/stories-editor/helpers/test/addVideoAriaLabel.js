/**
 * Internal dependencies
 */
import addVideoAriaLabel from '../addVideoAriaLabel';

describe( 'addVideoAriaLabel', () => {
	it( 'ignores arbitrary content', () => {
		const result = addVideoAriaLabel(
			<div>
				{ 'Ignore me' }
			</div>,
			{ name: 'anythin' },
			{},
		);

		expect( result ).toMatchSnapshot();
	} );

	it( 'ignores video content without an aria label attribute', () => {
		const result = addVideoAriaLabel(
			<div>
				{ 'Ignore me 2' }
			</div>,
			{ name: 'core/video' },
			{ ampAriaLabel: '' },
		);

		expect( result ).toMatchSnapshot();
	} );

	it( 'ignores video content if content is not a figure object', () => {
		const result = addVideoAriaLabel(
			<div>
				{ 'Ignore me 3' }
			</div>,
			{ name: 'core/video' },
			{ ampAriaLabel: 'Ignored label' },
		);

		expect( result ).toMatchSnapshot();
	} );

	it( 'ignores video content if content has incorrect node order', () => {
		const result = addVideoAriaLabel(
			<figure>
				<figcaption>
					{ 'Caption above content' }
				</figcaption>
				<video>
					{ 'Fallback content' }
				</video>
			</figure>,
			{ name: 'core/video' },
			{ ampAriaLabel: 'Ignored label' },
		);

		expect( result ).toMatchSnapshot();
	} );

	it( 'adds an aria label to video content', () => {
		const result = addVideoAriaLabel(
			<figure>
				<video>
					{ 'Fallback content' }
				</video>
			</figure>,
			{ name: 'core/video' },
			{ ampAriaLabel: 'Aria label 1' },
		);

		expect( result ).toMatchSnapshot();
	} );

	it( 'adds an aria label to amp-video content', () => {
		const result = addVideoAriaLabel(
			<figure>
				<amp-video>
					{ 'Fallback content' }
				</amp-video>
			</figure>,
			{ name: 'core/video' },
			{ ampAriaLabel: 'Aria label 1' },
		);

		expect( result ).toMatchSnapshot();
	} );

	it( 'adds an aria label to video content with a caption', () => {
		const result = addVideoAriaLabel(
			<figure>
				<video>
					{ 'Fallback content' }
				</video>
				<figcaption>
					{ 'Caption below content' }
				</figcaption>
			</figure>,
			{ name: 'core/video' },
			{ ampAriaLabel: 'Aria label 2' },
		);

		expect( result ).toMatchSnapshot();
	} );
} );
