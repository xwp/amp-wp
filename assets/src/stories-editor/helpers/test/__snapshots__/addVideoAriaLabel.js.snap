// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addVideoAriaLabel adds an aria label to amp-video content 1`] = `
<figure>
  <amp-video
    aria-label="Aria label 1"
  >
    Fallback content
  </amp-video>
</figure>
`;

exports[`addVideoAriaLabel adds an aria label to video content 1`] = `
<figure>
  <video
    aria-label="Aria label 1"
  >
    Fallback content
  </video>
</figure>
`;

exports[`addVideoAriaLabel adds an aria label to video content with a caption 1`] = `
<figure>
  <video
    aria-label="Aria label 2"
  >
    Fallback content
  </video>
  <figcaption>
    Caption below content
  </figcaption>
</figure>
`;

exports[`addVideoAriaLabel ignores arbitrary content 1`] = `
<div>
  Ignore me
</div>
`;

exports[`addVideoAriaLabel ignores video content if content has incorrect node order 1`] = `
<figure>
  <figcaption>
    Caption above content
  </figcaption>
  <video>
    Fallback content
  </video>
</figure>
`;

exports[`addVideoAriaLabel ignores video content if content is not a figure object 1`] = `
<div>
  Ignore me 3
</div>
`;

exports[`addVideoAriaLabel ignores video content without an aria label attribute 1`] = `
<div>
  Ignore me 2
</div>
`;
